platform/engineering/design-docs/2020-04-09-separate-content-build.md

```diff
@@ -0,0 +1,498 @@
+      page in Drupal or switching the flag in the JSON file
+1. Deploy a new content build
+
+## Specifics
+
+### Detailed Design
+
+#### Content build in a separate repository
+The content build will live in a separate repository called `content-build`.
+This repository will contain only code pertaining to the content build.
+
+![Separated content code base](images/separated-content-build/phase-2-separated-content-build.png)
+
+#### Triggering builds
+**Applications**  
+The `vets-website` build is triggered in CircleCI when:
+1. Code is pushed to a feature branch on GitHub
+1. A feature branch is merged into `master`
+1. A weekdaily deploy is triggered
+    - This build results in a versioned release artifact
+
+**Content**  
+The content build is triggered in Jenkins when an editor with the required
+permissions initiates a content deploy.
+
+#### Coordinating deployments to S3
+We use S3 to serve a static site. To do this, the bucket consists of a file tree
+which corresponds to the URLs on VA.gov. For example, `/education/index.html` is
+the file that's served at www.va.gov/education/.
+
+The output of the application build **currently** live in the `/generated/`
+directory. To support the transition, the **new** application build will put its
+assets in the `/applications/` directory.
+
+![Coordinating deployments to S3](images/separated-content-build/coordinating-deployments-to-s3.png)
+
+To ensure that one build won't override files in the other, we'll have to
+1. Update
+   [`vets-website-deploy.sh`](https://github.com/department-of-veterans-affairs/devops/blob/2c63e6d62fa5ad2b6b87478e30d0a7bc2820c252/ansible/deployment/roles/deploy-vets-website/files/vets-website-deploy.sh)
+   to copy only to the `/applications/` directory
+1. Create a new deploy script for the content deployment which copies everything
+   over _except_ the `/applications/` directory
+
+#### Creating temporary static pages for application testing
+- The Webpack configuration will use
+  [`html-webpack-plugin`](https://github.com/jantimon/html-webpack-plugin) to
+  generate temporary landing pages for React applications
+  - This will happen for all build types
+    - Engineers will be able to access the site on a `localhost` build
+    - CI can run E2E tests against production code without the output of the
+      content build
+- The E2E test script in CI will start a process to serve these pages so the
+  tests can run a browser and access the applications
+- **Important note:** These HTML files will not be served in production; the
+  landing pages for the applications come from the content build
+  - To ensure these pages aren't made in production, the deploy script will copy
+    only the files in `/applications/`
+
+#### Miscellaneous
+- Separating the content and application builds will require a standalone content validation job that runs outside of the CI pipeline- probably once a day on a schedule
```

Comment:
We'd want to run this before a deploy of the content, right? Otherwise we could deploy a set of HTML that references non-existence files. That's what this is trying to detect, right?

Reply:
Ideally, we'd run this before a content deploy, yes. Ideally, it'd block a deploy so we wouldn't have "broken" content make it to production. That's the ideal veteran experience, I'd say.

But because of reasons, I think this isn't feasible. Because we'll have hundreds of content editors, we can't block the deploy of _all_ content if _one_ thing (or even some higher threshold) is bad. We've talked at length about this with the CMS team, and the priority we settled on was build and deploy speed, which means all validation has to be taken out of that process, unfortunately.

So this is the next best approach, I think. I hope.

For what it's worth, the current content-only deploy does no content validation. That's all done in the full build. That we can push out "broken" content won't be new with this approach.